src/nvim/api/win_config.c

```diff
@@ -190,22 +216,71 @@ Window nvim_open_win(Buffer buffer, Boolean enter, Dict(float_config) *config, E
   if (!parse_float_config(config, &fconfig, false, true, err)) {
     return 0;
   }
-  win_T *wp = win_new_float(NULL, false, fconfig, err);
+
+  bool is_split = HAS_KEY(config, float_config, split) || HAS_KEY(config, float_config, vertical);
+
+  win_T *wp = NULL;
+  tabpage_T *tp = curtab;
+  if (is_split) {
+    win_T *parent = NULL;
+    if (!HAS_KEY(config, float_config, win) || config->win != -1) {
+      parent = find_window_by_handle(fconfig.window, err);
+      if (!parent) {
+        // find_window_by_handle has already set the error
+        return 0;
+      } else if (parent->w_floating) {
+        api_set_error(err, kErrorTypeException, "Cannot split a floating window");
+        return 0;
+      }
+    }
+
+    if (HAS_KEY(config, float_config, vertical) && !HAS_KEY(config, float_config, split)) {
+      if (config->vertical) {
+        fconfig.split = p_spr ? kWinSplitRight : kWinSplitLeft;
+      } else {
+        fconfig.split = p_sb ? kWinSplitBelow : kWinSplitAbove;
+      }
+    }
+    int flags = win_split_flags(fconfig.split, parent == NULL) | WSP_NOENTER;
+
+    if (parent == NULL) {
+      wp = win_split_ins(0, flags, NULL, 0);
+    } else {
+      tp = win_find_tabpage(parent);
+      switchwin_T switchwin;
+      // `parent` is valid in `tp`, so switch_win should not fail.
+      const int result = switch_win(&switchwin, parent, tp, true);
+      (void)result;
+      assert(result == OK);
+      wp = win_split_ins(0, flags, NULL, 0);
+      restore_win(&switchwin, true);
+    }
+    if (wp) {
+      wp->w_float_config = fconfig;
+    }
+  } else {
+    wp = win_new_float(NULL, false, fconfig, err);
+  }
   if (!wp) {
+    api_set_error(err, kErrorTypeException, "Failed to create window");
     return 0;
   }
-  if (enter) {
-    win_enter(wp, false);
+  switchwin_T switchwin;
+  if (switch_win_noblock(&switchwin, wp, tp, true) == OK) {
+    apply_autocmds(EVENT_WINNEW, NULL, NULL, false, curbuf);
   }
-  // autocmds in win_enter or win_set_buf below may close the window
-  if (win_valid(wp) && buffer > 0) {
-    Boolean noautocmd = !enter || fconfig.noautocmd;
-    win_set_buf(wp, buf, noautocmd, err);
-    if (!fconfig.noautocmd) {
-      apply_autocmds(EVENT_WINNEW, NULL, NULL, false, buf);
+  restore_win_noblock(&switchwin, true);
+  if (enter) {
+    goto_tabpage_win(tp, wp);
+    if (tabpage_win_valid(tp, wp) && curwin != wp) {
+      api_set_error(err, kErrorTypeException, "Failed to enter tabpage");
+      return 0;
     }
```

Comment:
It's just ensuring that the window was actually moved to the other tabpage if we're opening a window in another tabpage. Maybe it could be an assert instead since it's mainly a bug check and not a necessary runtime check?

Reply:
No, there can be autocommands that do anything while entering the window.

Reply:
The check below that on line 283 checks for the window being closed immediately, so we probably don't need to do both

Reply:
223d6060 thoughts?

Reply:
Ah actually nevermind about the assert. Not sure what I was thinking, don't want to crash nvim just because it's a debug build.